linked_list/linked_list.py

```diff
@@ -9,67 +9,167 @@ def __init__(self, value, next_node = None):
+            count = count + 1
+            current_node = current_node.next
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        count = 0
+
+        while current_node != None:
+            if count == index:
+                return current_node.value
+
+            count += 1
+            current_node = current_node.next
+        
+        return None
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_last(self):
-        pass
+        
+        #make sure list is not empty
+        if self.head == None:
+            return None
+
+        current_node = self.head
+        while current_node.next != None:
+            current_node = current_node.next
+        
+        return current_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head == None: 
+            self.head = new_node
+            return 
+        else:
+            current_node = self.head
+
+            while current_node.next != None:
+                current_node = current_node.next
+                
+            current_node.next = new_node 
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        
+        current_node = self.head
+
+        if self.head == None: 
+            return None
+
+        list_values = []
+        max_val = None
+        while current_node != None:
+            list_values.append(current_node.value)
+            current_node = current_node.next
+
+        max_val = max(list_values)
+        return max_val
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        temp = self.head
+        prev = self.head
+
+        if self.head == None: 
+            return None
+
+        if temp.value == value:
+            if temp.next is None:
+                return None
+            else:
+                temp.value = temp.next.value
+                temp.next = temp.next.next

```

Comment:
This doesn't seem like it should work, for one if the head is being deleted and there is no node behind it, nothing changes `self.head`.

```suggestion
        if self.head.value == value:
            self.head = self.head.next
            return
```